scipy/stats/_stats_py.py

```diff
@@ -342,18 +351,20 @@ def hmean(a, axis=0, dtype=None):
     """
     if not isinstance(a, np.ndarray):
         a = np.array(a, dtype=dtype)
-    if np.all(a >= 0):
-        # Harmonic mean only defined if greater than or equal to to zero.
+    elif dtype:
+        # Must change the default dtype allowing array type
         if isinstance(a, np.ma.MaskedArray):
-            size = a.count(axis)
+            a = np.ma.asarray(a, dtype=dtype)
         else:
-            if axis is None:
-                a = a.ravel()
-                size = a.shape[0]
-            else:
-                size = a.shape[axis]
+            a = np.asarray(a, dtype=dtype)
+
+    if np.all(a >= 0):
+        # Harmonic mean only defined if greater than or equal to zero.
+        if weights is not None:
+            weights = np.asanyarray(weights, dtype=dtype)
```

Comment:
I think we should do some input validation and check that the shape is coherent.

Reply:
Sanity checks are done in `np.average`, like in `gmean`.

https://github.com/numpy/numpy/blob/v1.22.0/numpy/lib/function_base.py#L395-L534

Reply:
Ok to omit it for this case, but only because the error message would use the same variable names. Still the traceback will be misleading as coming from NumPy instead of SciPy.

The point of validation here is to provide a help for the user on an error. In any case the code would fail, it's only a matter of failing in a controlled way for our use case.